base/strings/basic.jl

```diff
@@ -1,60 +1,210 @@
 
-string() = ""
-string(s::AbstractString) = s
+[self-synchronizing]: https://en.wikipedia.org/wiki/Self-synchronizing_code
 
-(::Type{Vector{UInt8}})(s::AbstractString) = Vector{UInt8}(String(s))
-(::Type{Array{UInt8}})(s::AbstractString) = Vector{UInt8}(s)
-(::Type{Vector{Char}})(s::AbstractString) = collect(s)
+Some string functions that extract code units, characters or substrings from
+strings error if you pass them out-of-bounds or invalid string indices. This
+includes `codeunit(s, i)`, `s[i]`, and `next(s, i)`. Functions that do string
+index arithmetic take a more relaxed approach to indexing and give you the
+closest valid string index when in-bounds, or when out-of-bounds, behave as if
+there were an infinite number of characters padding each side of the string.
+Usually these imaginary padding characters have code unit length `1` but string
+types may choose different "imaginary" character sizes as makes sense for their
+implementations (e.g. substrings may pass index arithmetic through to the
+underlying string they provide a view into). Relaxed indexing functions include
+those intended for index arithmetic: `thisind`, `nextind` and `prevind`. This
+model allows index arithmetic to work with out-of- bounds indices as
+intermediate values so long as one never uses them to retrieve a character,
+which often helps avoid needing to code around edge cases.
 
-Symbol(s::AbstractString) = Symbol(String(s))
+See also: [`codeunit`](@ref), [`ncodeunits`](@ref), [`thisind`](@ref),
+[`nextind`](@ref), [`prevind`](@ref)
+"""
+AbstractString
 
-# string types are convertible
-convert(::Type{T}, s::T) where {T<:AbstractString} = s
-convert(::Type{T}, s::AbstractString) where {T<:AbstractString} = T(s)
+## required string functions ##
 
-## generic supplied functions ##
+"""
+    ncodeunits(s::AbstractString) -> Int
 
-start(s::AbstractString) = 1
-done(s::AbstractString,i) = (i > endof(s))
-getindex(s::AbstractString, i::Int) = next(s,i)[1]
-getindex(s::AbstractString, i::Integer) = s[Int(i)]
-getindex(s::AbstractString, i::Colon) = s
-getindex(s::AbstractString, r::UnitRange{<:Integer}) = s[Int(first(r)):Int(last(r))]
-# TODO: handle other ranges with stride ±1 specially?
-getindex(s::AbstractString, v::AbstractVector{<:Integer}) =
-    sprint(length(v), io->(for i in v; write(io,s[i]) end))
-getindex(s::AbstractString, v::AbstractVector{Bool}) =
-    throw(ArgumentError("logical indexing not supported for strings"))
+Return the number of code units in a string. Indices that are in bounds to
+access this string must satisfy `1 ≤ i ≤ ncodeunits(s)`. Not all such indices
+are valid – they may not be the start of a character, but they will return a
+code unit value when calling `codeunit(s,i)`.
+
+See also: [`codeunit`](@ref), [`checkbounds`](@ref), [`sizeof`](@ref),
+[`length`](@ref), [`endof`](@ref)
+"""
+ncodeunits(s::AbstractString)
+
+"""
+    codeunit(s::AbstractString) -> Type{<:Union{UInt8, UInt16, UInt32}}
```

Comment:
(from a [discussion](https://discourse.julialang.org/t/changes-to-the-representation-of-char/7291/9?u=tamas_papp) on Discourse)

I think that the function that retrieves the type should perhaps have a different name than the function that returns a codeunit. eg `codeunit_type`. Methods of the same function usually share behavior on some abstract level, this is stretching that. I realize that the name is long, but it is only used in a few places (mostly to define the method for various types). I expect that in code outside `Base` it will be only used to help type inference so that this will not be cumbersome.